src/mjolnir/ferry_connections.cc

```diff
@@ -162,9 +163,12 @@ uint32_t ShortestPath(const uint32_t start_node_idx,
           continue;
         }
 
-        // Get cost - need the length and speed of the edge; Use a penalty if an edge is
-        // destination_only and calculate it in the cost
-        float penalty = w.destination_only() ? 300 : 0;
+        // if the first edge was destonly and the previous edge was as well, add no penalty
+        float penalty =
+            (first_edge_destonly ? (!pred_destonly && w.destination_only()) : w.destination_only())
+                ? 300
+                : 0;
```

Comment:
add a penalty only if destonly should be removed AND the pred was not destonly, or if the connecting edge wasn't even destonly